numba/tests/test_np_functions.py

```diff
@@ -4216,6 +4216,97 @@ def test_swapaxes_exception(self):
         self.assertIn('The third argument "axis2" is out of bounds for array'
                       ' of given dimension', str(raises.exception))
 
+    def test_take_along_axis(self):
+        a = np.arange(24).reshape((3, 1, 4, 2))
+
+        # For now axis must be literal, so we need to define a bunch of
+        # functions.
+        @njit
+        def axis_none(a, i):
+            return np.take_along_axis(a, i, axis=None)
+
+        indices = np.array([1, 2], dtype=np.uint64)
+        self.assertPreciseEqual(axis_none(a, indices),
+                                axis_none.py_func(a, indices))
+
+        ai0 = np.argsort(a, axis=0)
+        ai1 = np.argsort(a, axis=1)
+        ai2 = np.argsort(a, axis=2)
+        ai3 = np.argsort(a, axis=3)
+
+        @njit
+        def axis_zero(a, i):
+            return np.take_along_axis(a, i, axis=0)
+
+        self.assertPreciseEqual(axis_zero(a, ai0), axis_zero.py_func(a, ai0))
+
+        @njit
+        def axis_one(a, i):
+            return np.take_along_axis(a, i, axis=1)
+
+        self.assertPreciseEqual(axis_one(a, ai1), axis_one.py_func(a, ai1))
+
+        @njit
+        def axis_two(a, i):
+            return np.take_along_axis(a, i, axis=2)
+
+        self.assertPreciseEqual(axis_two(a, ai2), axis_two.py_func(a, ai2))
+
+        @njit
+        def axis_three(a, i):
+            return np.take_along_axis(a, i, axis=3)
+
+        self.assertPreciseEqual(axis_three(a, ai3), axis_three.py_func(a, ai3))
+
+        @njit
+        def axis_minus_one(a, i):
+            return np.take_along_axis(a, i, axis=-1)
+
+        self.assertPreciseEqual(axis_minus_one(a, ai3),
+                                axis_minus_one.py_func(a, ai3))
+
```

Comment:
This:
```python
from numba import njit
import numpy as np

# Based on https://github.com/numpy/numpy/blob/v1.21.0/numpy/lib/tests/test_shape_base.py#L74-L79
a  = np.ones((3, 4, 1))
ai = np.ones((1, 2, 5), dtype=np.intp)

@njit
def check(a, i):
    return np.take_along_axis(a, i, axis=1)

actual = check(a, ai)
assert actual.shape == (3, 2, 5)
```
Seems to segfault locally.

Reply:
```
Program received signal SIGSEGV, Segmentation fault.
0x00007fffd53f0111 in numba::np::arrayobj::_take_along_axis_impl$244 () at arrayobj.py:5482
5482                        out_1d[j] = a_1d[indices_1d[j]]
```

Reply:
```
$ NUMBA_BOUNDSCHECK=1 python test_7202.py
debug: IndexError: index 1 is out of bounds for axis 0 with size 1
Traceback (most recent call last):
  File "test_7202.py", line 12, in <module>
    actual = check(a, ai)
IndexError: index is out of bounds
```
Looks like it's an invalid array access causing this.

Reply:
Turns out the alternative `take_along_axis` implementation in NumPy docs (which this PR is based on) doesn't match real implementation in some edge cases. I filed https://github.com/numpy/numpy/issues/19704

Reply:
OK, thanks to @eric-wieser I think I can make this work in regular Python... but Numba is missing `np.broadcast_to()` (though there's a private function?) and is also missing the more complex `nditer()` options. So next is seeing if I can implement matching code some other way that Numba does support.

Reply:
Numba now has `broadcast_to`: #7119.

Reply:
OK, done, but you should check my logic.